File: basic/search/binary_search_2.py
class Binary_Search_2(object):

    def __init__(self) -> None:
        super().__init__()

    def search(self, arr, item):
        '''查找最后一个值等于给定值的元素'''
        l = 0
        r = len(arr) - 1
        while True:
            m = (l + r) // 2
            if arr[m] < item:
                l = m + 1
            elif arr[m] > item:
                r = m - 1
            else:
                if m == len(arr) - 1 or arr[m + 1] != item:  # 终止条件
                    return m
                l = m + 1
            if l > r:  #注意二分查找可以相等,与排序的递归退出条件不同
                return -1


if __name__ == '__main__':
    arr = [int(x) for x in input().strip().split(' ')]
    item = int(input().strip())
    binary_search = Binary_Search_2()
    idx = binary_search.search(arr, item)
    print(idx)

File: basic/search/binary_search_2.cpp
#include <iostream>

class Solution {
 public:
  int binary_search(int arr[], int len, int key) {
    int l = 0, r = len - 1;
    while (l <= r) {  // <=
      int m = l + (r - l) / 2;
      if (arr[m] > key) {
        r = m - 1;
      } else if (arr[m] < key) {
        l = m + 1;
      } else {
        if (m + 1 < len && arr[m] != arr[m + 1]) {
          return m;
        }
        l = m + 1;
      }
    }
    return -1;
  }

  int binary_search2(int arr[], int len, int key) {
    int l = 0, r = len - 1;
    while (l < r) {                 // <
      int m = l + (r - l + 1) / 2;  // 向上取整
      if (arr[m] > key) {
        r = m - 1;
      } else {  // 找最后一个<=key, 主要是上一个方法不能应对需要查找元素不相邻的情况
        l = m;
      }
    }
    return (arr[l] == key) ? l : -1;
  }
};

int main() {
  int a[6] = {1, 2, 3, 5, 5, 7};

  Solution solution = Solution();
  int idx = solution.binary_search(a, 6, 5);
  std::cout << "idx: " << idx << std::endl;
  int idx2 = solution.binary_search2(a, 6, 5);
  std::cout << "idx2: " << idx2 << std::endl;
}